### DISCUSSION TITLE
Fan read / control on IdeaPads

### DISCUSSION THREAD

**Discussion by kasti0** [zhen-zen/YogaSMC#65] [Ideas]
i tried to do it myself on windows with NBFanControl with probing the EC and examining my DSDT.
however I still couldn´t find the EC registers or methods that are related to the laptop fans.
my laptop is an IdeaPad 330s 15ik 81F5

**Comment by zhen-zen** [comment]
It's not implemented yet since mine don't support that.

**Comment by marianopela** [comment]
Is there anything I can do to help? I think mine supports it, as in the advanced bios there are many fan control settings

**Comment by zhen-zen** [comment]
Some models have it exposed directly to EC, such as NBFC ones. And there are also `VPCCMD` commands. https://github.com/zhen-zen/YogaSMC/blob/master/YogaSMC/YogaVPC/IdeaVPC.hpp#L50

**Comment by marianopela** [comment]
Would this be something useful?

```
[    5.097431]: Clearing out battery data
[    5.097518]: SMCGetKeyInfo cmd 0x9(key     ) size 0x2
[    5.097568]: SMCGetKeyInfo cmd 0xa(key     ) size 0x2
[    5.097605]: SMC key      for cmd 0xb is not supported
[    5.097653]: SMC key      for cmd 0x76 is not supported
[    5.097692]: SMC key      for cmd 0x2 is not supported
[    5.097727]: SMC key      for cmd 0x1 is not supported
[    5.097774]: SMCGetKeyInfo cmd 0x16(key     ) size 0x2
[    5.097824]: SMC key      for cmd 0x54 is not supported
[    5.097870]: SMC key      for cmd 0x8 is not supported
[    5.097913]: SMCGetKeyInfo cmd 0x17(key     ) size 0x2
[    5.097956]: SMCGetKeyInfo cmd 0x13(key     ) size 0x2
[    5.097999]: SMCGetKeyInfo cmd 0x53(key     ) size 0x2
[    5.098043]: SMC key      for cmd 0x9c is not supported
[    5.098087]: SMC key      for cmd 0x12 is not supported
[    5.098131]: SMC key      for cmd 0x8b is not supported
[    5.098178]: SMC key      for cmd 0x21 is not supported
[    5.098222]: SMCGetKeyInfo cmd 0xf(key     ) size 0x2
[    5.098268]: SMCGetKeyInfo cmd 0x10(key     ) size 0x2
[    5.098312]: SMC key      for cmd 0x18 is not supported
```

**Comment by zhen-zen** [comment]
No, they are not published as SMC keys. There are few documents on it, so I haven't implement that.

**Comment by zhen-zen** [comment]
The fan mode control should be available with `VPCCMD_R_FAN` and `VPCCMD_W_FAN ` though the `VPCR`/`VPCW` interface on certain models.

To read it manually:
1. Allow direct EC access
`ioio -s IdeaVPC ECLock false`
2. Read `VPCCMD_R_FAN`
`ioio -s IdeaVPC ReadEC 0x2B`


Specific fan speed RPM might be available in EC regions, which could be identified by putting the laptop under high load from sleeping (speed = 0). Presets from https://github.com/hirschmann/nbfc are available for certain models.

**Reply by marianopela** [reply]
Hi, thanks for the Discussions section, I really like it!

> To read it manually:
> 
> 1. Allow direct EC access
>    `ioio -s IdeaVPC ECLock false`
> 2. Read `VPCCMD_R_FAN`
>    `ioio -s IdeaVPC ReadEC 0x2B`

I was wondering, what should we expect from these commands, and how do we know if it is supported? I'm still struggling to understand how EC reading works
![Screenshot 2020-12-12 at 12 31 00](https://user-images.githubusercontent.com/42821024/101982728-30f37080-3c76-11eb-979e-9fb36315fd15.png)

**Reply by zhen-zen** [reply]
The result will be shown in `dmesg` log as well. It's a pity that system log doesn't work on Big Sur, which could watch real-time log.

**Reply by marianopela** [reply]
You're right. By continuously switching between systems I forgot about this

```
[    5.722588]: YSMC - Info: IdeaWMI::WMIY Unknown yoga mode: 0
[   38.329499]: failed to write xattr for NSIRD_airportd_IVuM7o: 2failed to write xattr for NSIRD_securityd_tnA7Vy: 2Sandbox: 2 duplicate reports for imagent deny(1) mach-lookup com.apple.backupd.xpcSandbox: taskgated-helper(440) deny(1) user-preference-read kCFPreferencesAnyApplicationYSMC - Info: YogaSMCUserClient::IdeaVPC registerNotificationPort subscribed
[  140.806336]: YSMC - Info: IdeaVPC::VPC0 ReadEC timeout 0x1a
[  140.806354]: YSMC - Info: IdeaVPC::VPC0 Failed to read VPC 34
[  167.962799]: Waking up reference: 41Thread waiting on reference 41 woke upevaluation result: 41, allowed, cache,  1607691387YSMC - Info: IdeaVPC::VPC0 direct VPC EC manipulation is unlocked
[  180.478433]: YSMC - Info: IdeaVPC::VPC0 ReadEC timeout 0x2b
[  180.478457]: YSMC - Info: IdeaVPC::VPC0 ReadEC failed 0x2b 42
[  259.221404]: YSMC - Info: IdeaVPC::VPC0 ReadEC 0x2b result: 0x0 0
[  331.363393]: YSMC - Info: IdeaVPC::VPC0 ReadEC timeout 0x2b
[  331.363419]: YSMC - Info: IdeaVPC::VPC0 ReadEC failed 0x2b 14
[  402.461605]: YSMC - Info: IdeaVPC::VPC0 ReadEC timeout 0x2b
[  402.461632]: YSMC - Info: IdeaVPC::VPC0 ReadEC failed 0x2b 5
```

**Reply by zhen-zen** [reply]
It's hard to say if that feature also suffers from EC evaluation failure. However, the result doesn't seem promising for now. Do you have fan related control in Windows with vantage?

**Comment by kasti0** [comment]
for the ideapad 330s 15ikb I don´t have any fan control with lenovo vantage.

**Comment by roneyrogerio** [comment]
This may help in something, disregard if it is irrelevant:
https://www.youtube.com/watch?v=zjw93YkpHhU
https://gitlab.com/OdinTdh/extremecooling4linux/-/blob/master/ec4Linux.py
https://github.com/hirschmann/nbfc/issues/526
